presto-spi/src/main/java/com/facebook/presto/spi/block/MaskedBlock.java

```diff
@@ -0,0 +1,210 @@
+    {
+        block.writeBytesTo(mapPosition(position), offset, length, blockBuilder);
+    }
+
+    @Override
+    public void writePositionTo(int position, BlockBuilder blockBuilder)
+    {
+        block.writePositionTo(mapPosition(position), blockBuilder);
+    }
+
+    @Override
+    public boolean equals(int position, int offset, Block otherBlock, int otherPosition, int otherOffset, int length)
+    {
+        return block.equals(mapPosition(position), offset, otherBlock, otherPosition, otherOffset, length);
+    }
+
+    @Override
+    public long hash(int position, int offset, int length)
+    {
+        return block.hash(mapPosition(position), offset, length);
+    }
+
+    @Override
+    public int compareTo(int leftPosition, int leftOffset, int leftLength, Block rightBlock, int rightPosition, int rightOffset, int rightLength)
+    {
+        return block.compareTo(mapPosition(leftPosition), leftOffset, leftLength, rightBlock, rightPosition, rightOffset, rightLength);
+    }
+
+    @Override
+    public Block getSingleValueBlock(int position)
+    {
+        return block.getSingleValueBlock(mapPosition(position));
+    }
+
+    @Override
+    public int getPositionCount()
+    {
+        return mappingLength;
+    }
+
+    @Override
+    public int getSizeInBytes()
+    {
+        return block.getSizeInBytes();
+    }
+
+    @Override
+    public int getRetainedSizeInBytes()
+    {
+        return block.getRetainedSizeInBytes();
+    }
+
+    @Override
+    public BlockEncoding getEncoding()
+    {
+        return new MaskedBlockEncoding(block.getEncoding());
+    }
+
+    @Override
+    public Block copyPositions(List<Integer> positions)
+    {
+        List<Integer> mappedPositions = positions.stream().map(this::mapPosition).collect(toList());
+        return block.copyPositions(mappedPositions);
+    }
+
+    @Override
+    public Block getRegion(int positionOffset, int length)
+    {
+        return new MaskedBlock(block, positionsMapping, mappingOffset + positionOffset, length);
+    }
+
+    @Override
+    public Block copyRegion(int startPosition, int length)
+    {
+        List<Integer> mappedPositions = new ArrayList<>(length);
+        for (int position = startPosition; position < startPosition + length; position++) {
+            mappedPositions.add(mapPosition(position));
+        }
+        return block.copyPositions(mappedPositions);
+    }
+
+    @Override
+    public boolean isNull(int position)
+    {
+        return block.isNull(mapPosition(position));
+    }
+
+    @Override
+    public void assureLoaded()
+    {
+        block.assureLoaded();
+    }
+
+    private int mapPosition(int position)
+    {
+        checkReadablePosition(position);
+        return positionsMapping[position + mappingOffset];
+    }
+
+    private void checkReadablePosition(int position)
+    {
+        if (position < 0 || position >= getPositionCount()) {
+            throw new IllegalArgumentException("position is not valid");
```

Comment:
this throws also for other cases, like negative position or position way to big (larger then masked block)